x-pack/legacy/plugins/alerting/server/plugin.ts

```diff
@@ -0,0 +1,168 @@
+          },
+          ui: [],
+          api: ['alerting-read'],
+        },
+      },
+    });
+
+    // Encrypted attributes
+    plugins.encrypted_saved_objects.registerType({
+      type: 'alert',
+      attributesToEncrypt: new Set(['apiKey']),
+      attributesToExcludeFromAAD: new Set([
+        'scheduledTaskId',
+        'muted',
+        'mutedInstanceIds',
+        'updatedBy',
+      ]),
+    });
+
+    const alertTypeRegistry = new AlertTypeRegistry({
+      taskManager: plugins.task_manager,
+      taskRunnerFactory: this.taskRunnerFactory,
+    });
+    this.alertTypeRegistry = alertTypeRegistry;
+    this.serverBasePath = core.http.basePath.serverBasePath;
+
+    // Register routes
+    core.http.route(createAlertRoute);
+    core.http.route(deleteAlertRoute);
+    core.http.route(findAlertRoute);
+    core.http.route(getAlertRoute);
+    core.http.route(listAlertTypesRoute);
+    core.http.route(updateAlertRoute);
+    core.http.route(enableAlertRoute);
+    core.http.route(disableAlertRoute);
+    core.http.route(updateApiKeyRoute);
+    core.http.route(muteAllAlertRoute);
+    core.http.route(unmuteAllAlertRoute);
+    core.http.route(muteAlertInstanceRoute);
+    core.http.route(unmuteAlertInstanceRoute);
+
+    return {
+      registerType: alertTypeRegistry.register.bind(alertTypeRegistry),
+    };
+  }
+
+  public start(core: AlertingCoreStart, plugins: AlertingPluginsStart): PluginStartContract {
+    const { adminClient, serverBasePath } = this;
+
+    const alertsClientFactory = new AlertsClientFactory({
+      alertTypeRegistry: this.alertTypeRegistry!,
+      logger: this.logger,
+      taskManager: plugins.task_manager,
+      securityPluginSetup: plugins.security,
+      getSpaceId(request: Hapi.Request) {
+        const spacesPlugin = plugins.spaces();
+        return spacesPlugin ? spacesPlugin.getSpaceId(request) : undefined;
+      },
+    });
+
+    this.taskRunnerFactory.initialize({
+      logger: this.logger,
+      isSecurityEnabled: !!plugins.security,
+      getServices(request: Hapi.Request): Services {
+        return {
+          callCluster: (...args) =>
+            adminClient!.asScoped(KibanaRequest.from(request)).callAsCurrentUser(...args),
+          savedObjectsClient: core.savedObjects.getScopedSavedObjectsClient(request),
+        };
+      },
+      executeAction: plugins.actions.execute,
+      encryptedSavedObjectsPlugin: plugins.encrypted_saved_objects,
+      spaceIdToNamespace(spaceId?: string): string | undefined {
+        const spacesPlugin = plugins.spaces();
+        return spacesPlugin && spaceId ? spacesPlugin.spaceIdToNamespace(spaceId) : undefined;
+      },
+      getBasePath(spaceId?: string): string {
+        const spacesPlugin = plugins.spaces();
+        return spacesPlugin && spaceId ? spacesPlugin.getBasePath(spaceId) : serverBasePath!;
+      },
+    });
+
+    return {
+      listTypes: this.alertTypeRegistry!.list.bind(this.alertTypeRegistry!),
```

Comment:
seems like perhaps `listTypes` should be on `setup` and not `start`; seems weird you can write to the registry at setup but not read it?  Perhaps it should be in both `setup` AND `start`, not sure if that's a common scenario tho.

Reply:
The way I saw it for `listTypes` is we don't have a full list until our plugin is started. By that point it would mean all the other plugins have registered their alert types and we can provide a non changing list of alert types within `start`.

If we expose it on `setup`, those who use the API on setup would probably see the result vary by order of operations (1st plugin sees nothing, 2nd plugin sees alert types registered by 1st, etc). Right?